Adding a detailed json-schema

eneuhauser opened the pull request:
I saw that you just removed the json-schema, but I believe this schema is a good representation of JSON API and concisely explains the format.

tkellen commented:
Thanks for this @eneuhauser! I'm going to hold off on merging until we have marked 1.0.

amaltson commented:
With 1.0 released, will this be considered for merging? This would help when documenting APIs with RAML.

tkellen commented:
Yes! Can you squash this to a single commit please?

ethanresnick commented on schema:
```diff
@@ -0,0 +1,243 @@
+          "type": "string",
+          "format": "uri"
+        },
+        "linkage": {
```
This should be `data`

ethanresnick commented on schema:
```diff
@@ -0,0 +1,243 @@
+          "type": "string",
+          "format": "uri"
+        },
+        "linkage": {
```
ditto

ethanresnick commented on schema:
```diff
@@ -0,0 +1,243 @@
+        "links": { "$ref": "#/definitions/links" },
+        "meta": { "$ref": "#/definitions/meta" }
+      },
+      "additionalProperties": true
```
There should be attributes and relationships definitions here, I think, and then additionalProperties should be false

ethanresnick commented on schema:
```diff
@@ -0,0 +1,243 @@
+          "uniqueItems": true
+        }
+      },
+      "additionalProperties": true
```
additionalProperties should be false here too, though meta is now allowed

slaskis commented:
I've been using this schema and updated my version of it now to pass with 1.0 (including your comments @ethanresnick).

Since you did most of the work feel free to update this PR with my gist: https://gist.github.com/slaskis/f4f32beeea0080320e00

eneuhauser commented:
@tkellen I've updated the schema to match the new 1.0 format and squashed into a single commit. I validated the schema complies with the example on the homepage using [JSON Schema Lint](http://jsonschemalint.com/draft4/#).

@slaskis, thanks for your help. I used your changes to start out, but I did make some modifications based on the current format of the spec. Please let me know if you see any issues.

slaskis commented:
@eneuhauser happy to help! 

I was a bit lazy with the descriptions but I see you took care of that :+1:

slaskis commented on schema:
```diff
@@ -0,0 +1,357 @@
+            "$ref": "#/definitions/resource"
+          },
+          "uniqueItems": true
+        }
```
Perhaps this could be moved into something like "#/definitions/collection"? Not that it's re-used right now but just for clarity. If that's even the correct term, but "an array of resource objects" doesn't really fit in a nice and tidy string.

eneuhauser replied:
I don't know about this one. You wouldn't create a new object to represent an array of resources, so I don't think it needs its own definition.

slaskis replied:
Very good points, ignore my ramblings please :)

slaskis commented on schema:
```diff
@@ -0,0 +1,357 @@
+			"description": "Members of the attributes object (\"attributes\") represent information about the resource object in which it's defined.",
+      "type": "object",
+			"patternProperties": {
+				"^\\w[\\w-_]*$": {
```
Is it too strict to filter out the [reserved member names](http://jsonapi.org/format/#document-resource-object-attributes) here? That is adding `relationships` and `links` to the regex.

eneuhauser replied:
I had originally tried this, but felt it was unnecessary; however, I never turn down a regex challenge, so I came up with the following: `^(?!relationships$)(?!links$)\\w[\\w-_]*$`.

slaskis replied:
Nice, regex golf time! What about merging those into `^(?!relationships|links$)\\w[\\w-_]*$`? ;)

tedepstein commented on faq/index.md:
```diff
@@ -66,7 +66,14 @@ alternative ways of requesting certain operations (e.g., resource creation via
+[http://jsonapi.org/schema](http://jsonapi.org/schema). Please note that this
+schema is not a perfect document. Just because a JSON document may validate
+against this schema, that does not necessarily mean it is a valid JSON API
+document. The schema is provided for a base level sanity check.
```
Following this PR, and the preceding issue #271, I agree that a schema is helpful.  If it doesn't validate all of the semantics of JSON API, then I would say it's like most schemas in that respect.  It can yield some "false positives," where a given JSON input passes schema validation, but breaks some JSON API rules. 

Still, I assume that whatever this schema does validate is in fact a requirement of the JSON API spec.  So it should not yield any "false negatives."  If a given input fails validation against this schema, then it's definitely not valid JSON API.

That's an important guarantee, and one worth stating in the FAQ, I think.  Rather than saying it's "not a perfect document," say something like "Please note that this schema only enforces a subset of the JSON API specification."

eneuhauser replied:
This wording was taken from the previous JSON Schema. The previous schema was a lot more simplistic, so your suggestions are more appropriate for this version of the schema.

I reused the wording because I didn't want to get into the politics of how it should read, but I can take a crack at updating the copy match what this schema does and doesn't do.

tkellen replied:
Would you mind doing that @eneuhauser and pinging back when ready? I'd love to get this merged but it would be nice for it to be a bit more clear.

Commit:
Updating schema to match 1.0 version of the spec and updating FAQ documentation.

eneuhauser commented:
@tkellen, I've updated the FAQ with the spirit of the schema definition. I also cleaned up some whitespace issues with the schema if you notice differences there. Please review and let me know if it is sufficient. Thanks!

tkellen commented:
thanks @eneuhauser!

cc @dgeb @ethanresnick @steveklabnik

dgeb commented:
I haven't reviewed the schema definition in detail, but it looks like a great start.

I suggest we merge and allow for refinement (if needed).

Thanks so much for your work here @eneuhauser!

bf4 commented:
Just want to confirm, unless I'm not seeing it, that this schema was manually generated and is not as authoritative as the `/format` page? e.g. it includes `info` which isn't there.

ethanresnick commented:
> Just want to confirm, unless I'm not seeing it, that this schema was manually generated and is not as authoritative as the /format page

Correct. The current schema is non-normative and has a few known issues (see #867)

bf4 commented:
ref multiple schemas PR for discussion https://github.com/json-api/json-api/pull/1127
